### DISCUSSION TITLE
Convert engine using PKA to provider

### DISCUSSION THREAD

**Discussion by darkage2** [openssl/openssl#22887] [Q&A]
Hi,

We're in the process of upgrading our 1.1.1w version to the 3.x on a PPC based ASIC board. The first difficulty is that is running ENEAs OSE which if I remember correctly is not officially supported. Due to this, we already add some modifications on top of the releases.

The board supports hardware based PKA which is used in the current implementations engine. The default RSA, DH and DSA method is duplicated and the bn_mod_exp function is set to use it. The engine init function just initializes the data vectors for the accelerator.

Since providers don't support this type of approach, what would be the best way to proceed with? It feels like a bit of an overkill just to create a whole provider for these small changes.
Should we just create a copy of the default provider and change them to our implementations?
Since we're already adding modifications to the original code, can't we just simply overwrite it there nad use our functions instead?

**Comment by petrovr** [comment]
The problem with existing OpenSSL 3.+ code base is that is possible key to lost information for associated method. 

----
Once a key related material is obtained first step is to set key-method. After this key could be used as before.

For instance  [set pkcs#11 key method immediately on X.509 certificate ](https://gitlab.com/secsh/pkixssh/-/commit/d84457fbfca237d40da42526e2a051209b5eeaa3) allows PKIX-SSH X.509 certificate obtained from secure token to work as before. 
Another work-around could be based on [throw out misfunctional OpenSSL 3+ "key manager" in pkcs#11 helper client ](https://gitlab.com/secsh/pkixssh/-/commit/1ef9e2b6f30b05b4bfcb59607e4c31ec460fceeb). Idea here is when key method is assigned to re-assign key then key and pkey could be used as before.
PKIX-SSH code has more commits that "set pkcs#11 ....key method immediately on plain key".

----
Let see [NSS RSA engine code ](https://gitlab.com/e_nss/e_nss/-/blob/master/engines/e_nss_rsa.c) .
The important point is that "key" must be recreated for 3.0 -  method nss_init_keyctx_rsa:

- rsa = RSA_new_method(e);
- /* copy public key */
- EVP_PKEY_set1_RSA(pkey, rsa)

Remark:  nss_init_keyctx_XXX() is called by nss_load_key() for XXX keys.

Also engines set flag ENGINE_FLAGS_NO_REGISTER_ALL (source in e_nss.c).

----

I expect if provider is not associated with key functionality remain as before.
